pkg/ingester/metrics.go

```diff
@@ -0,0 +1,186 @@
+			Help: "The total number of chunks returned from queries.",
+			// A small number of chunks per series - 10*(8^(7-1)) = 2.6m.
+			Buckets: prometheus.ExponentialBuckets(10, 8, 7),
+		}),
+	}
+
+	if r != nil {
+		r.MustRegister(
+			m.flushQueueLength,
+			m.ingestedSamples,
+			m.ingestedSamplesFail,
+			m.queries,
+			m.queriedSamples,
+			m.queriedSeries,
+			m.queriedChunks,
+		)
+	}
+
+	return m
+}
+
+// TSDB shipper metrics
+type shipperMetrics struct {
+	dirSyncs        *prometheus.Desc
+	dirSyncFailures *prometheus.Desc
+	uploads         *prometheus.Desc
+	uploadFailures  *prometheus.Desc
+
+	regsMu sync.RWMutex                    // custom mutex for shipper registry, to avoid blocking main user state mutex on collection
+	regs   map[string]*prometheus.Registry // One prometheus registry (used by shipper) per tenant
+}
+
+func newShipperMetrics(r prometheus.Registerer) *shipperMetrics {
+	m := &shipperMetrics{
+		regs: make(map[string]*prometheus.Registry),
+
+		dirSyncs: prometheus.NewDesc(
+			"cortex_ingester_shipper_dir_syncs_total",
+			"TSDB: Total dir sync attempts",
+			nil, nil),
+		dirSyncFailures: prometheus.NewDesc(
+			"cortex_ingester_shipper_dir_sync_failures_total",
+			"TSDB: Total number of failed dir syncs",
+			nil, nil),
+		uploads: prometheus.NewDesc(
+			"cortex_ingester_shipper_uploads_total",
+			"TSDB: Total object upload attempts",
+			nil, nil),
+		uploadFailures: prometheus.NewDesc(
+			"cortex_ingester_shipper_upload_failures_total",
+			"TSDB: Total number of failed object uploads",
+			nil, nil),
+	}
+
+	if r != nil {
+		r.MustRegister(m)
+	}
+	return m
+}
+
+func (sm *shipperMetrics) Describe(out chan<- *prometheus.Desc) {
+	out <- sm.dirSyncs
+	out <- sm.dirSyncFailures
+	out <- sm.uploads
+	out <- sm.uploadFailures
+}
+
+func (sm *shipperMetrics) Collect(out chan<- prometheus.Metric) {
+	gathered := make(map[string][]*dto.MetricFamily)
+
+	regs := sm.shipperRegistries()
+	for userID, r := range regs {
+		m, err := r.Gather()
+		if err != nil {
+			level.Debug(util.Logger).Log("msg", "failed to gather metrics from TSDB shipper", "user", userID, "err", err)
+			continue
+		}
+
+		addToGatheredMap(gathered, m)
+	}
+
+	// OK, we have it all. Let's build results.
+	out <- prometheus.MustNewConstMetric(sm.dirSyncs, prometheus.CounterValue, sumCounters(gathered["thanos_shipper_dir_syncs_total"]))
+	out <- prometheus.MustNewConstMetric(sm.dirSyncFailures, prometheus.CounterValue, sumCounters(gathered["thanos_shipper_dir_sync_failures_total"]))
+	out <- prometheus.MustNewConstMetric(sm.uploads, prometheus.CounterValue, sumCounters(gathered["thanos_shipper_uploads_total"]))
+	out <- prometheus.MustNewConstMetric(sm.uploadFailures, prometheus.CounterValue, sumCounters(gathered["thanos_shipper_upload_failures_total"]))
+}
+
+func (sm *shipperMetrics) shipperRegistries() []*prometheus.Registry {
+	sm.regsMu.RLock()
+	defer sm.regsMu.RUnlock()
+
+	regs := []*prometheus.Registry(nil)
```

Comment:
`nit` this looks weird to my eyes. I'd prefer `regs := []*prometheus.Registry{}`

Reply:
Actually is there a reason they're being copied out? Couldn't we just return `sm.regs` ?

Reply:
> Actually is there a reason they're being copied out? Couldn't we just return `sm.regs` ?

We would need to keep the lock while doing that. I don't like the idea of metrics code blocking ingestion path (when opening new TSDB), and opted for copying here, but perhaps it's not a big deal. Not sure.

When it comes to declaration, either `nil` version or `var regs []*prometheus.Registry` are fine, but `regs := []*prometheus.Registry{}` allocates non-nil zero-length slice, which is just useless allocation.

Reply:
I think keeping the current code - which reduce the time the lock is taken - is better.

Reply:
Okay, leave as is.

Reply:
One other possibility would be to cache the slice of registries, and nil it when adding new user. We can do that if making a copy for each metrics becomes a problem, which is unlikely.